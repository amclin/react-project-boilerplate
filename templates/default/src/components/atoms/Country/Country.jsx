import React from 'react'
import { useQuery } from '@apollo/react-hooks'
import PropTypes from 'prop-types'
import COUNTRY_QUERY from '../../../graphql/queries/country'

const Country = ({ code }) => {
  const { loading, error, data } = useQuery(COUNTRY_QUERY, {
    variables: { code }
  })

  if (!code) {
    return (
      <div data-testid="no-code-div" className="country">
        No country selected
      </div>
    )
  }
  if (error) {
    return (
      <div data-testid="error-div" className="country">
        Error!
      </div>
    )
  }
  if (loading) {
    return (
      <div data-testid="loading-div" className="country">
        Loading...
      </div>
    )
  }
  if (data.country) {
    const { name, emoji } = data.country

    return (
      <div className="country">
        <div data-testid="country-name">{name}</div>
        <div data-testid="country-emoji">{emoji}</div>
      </div>
    )
  }
  return (
    <div data-testid="no-data-div" className="country">
      No Country Data
    </div>
  )
}

Country.propTypes = {
  code: PropTypes.string
}

Country.defaultProps = {
  code: ''
}

export default Country
