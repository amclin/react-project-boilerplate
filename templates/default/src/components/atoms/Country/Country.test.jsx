import React from 'react'
import { render } from '@testing-library/react'
import { useQuery } from '@apollo/react-hooks'
import Country from '.'

jest.mock('@apollo/react-hooks')

describe('Country', () => {
  let mockQueryResponse = {}

  it('should render no country output when no code provided', () => {
    mockQueryResponse = {
      loading: true
    }
    useQuery.mockReturnValue(mockQueryResponse)
    const { getByTestId } = render(<Country />)

    expect(getByTestId('no-code-div').textContent).toBe('No country selected')
  })
  it('should render error output on error', () => {
    mockQueryResponse = {
      error: {
        graphQLErrors: []
      }
    }
    useQuery.mockReturnValue(mockQueryResponse)
    const { getByTestId } = render(<Country code="US" />)

    expect(getByTestId('error-div').textContent).toBe('Error!')
  })
  it('should render loading output when loading', () => {
    mockQueryResponse = {
      loading: true
    }
    useQuery.mockReturnValue(mockQueryResponse)
    const { getByTestId } = render(<Country code="US" />)

    expect(getByTestId('loading-div').textContent).toBe('Loading...')
  })

  it('should render empty country output when no data', () => {
    mockQueryResponse = {
      loading: false,
      data: {}
    }
    useQuery.mockReturnValue(mockQueryResponse)
    const { getByTestId } = render(<Country code="NONE" />)

    expect(getByTestId('no-data-div').textContent).toBe('No Country Data')
  })

  it('should render a Country', () => {
    mockQueryResponse = {
      loading: false,
      data: {
        country: {
          name: 'United States',
          code: 'US',
          emoji: 'E'
        }
      }
    }
    useQuery.mockReturnValue(mockQueryResponse)
    const { asFragment, getByTestId } = render(<Country code="US" />)
    const name = getByTestId('country-name')
    const emoji = getByTestId('country-emoji')

    expect(name.textContent).toBe(mockQueryResponse.data.country.name)
    expect(emoji.textContent).toBe(mockQueryResponse.data.country.emoji)
    expect(asFragment()).toMatchSnapshot()
  })
})
