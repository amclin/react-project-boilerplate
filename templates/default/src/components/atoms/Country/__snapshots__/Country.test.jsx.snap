// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Country should render a Country 1`] = `
<DocumentFragment>
  <div
    class="country"
  >
    <div
      data-testid="country-name"
    >
      United States
    </div>
    <div
      data-testid="country-emoji"
    >
      E
    </div>
  </div>
</DocumentFragment>
`;
